MSC4108: Mechanism to allow OAuth 2.0 API sign in and E2EE set up via QR code

hughns opened the pull request:
**SCT Stuff**

[MSC Checklist](https://github.com/matrix-org/matrix-spec-proposals/pull/4108#issuecomment-4500516017)

[FCP tickyboxes](https://github.com/matrix-org/matrix-spec-proposals/pull/4108#issuecomment-4500521390)

Designated reviewers:

* [ ] @erikjohnston 
* [x] @richvdh 

----

[Rendered](https://github.com/matrix-org/matrix-spec-proposals/blob/element-hq/oidc-qr-login/proposals/4108-oidc-qr-login.md)

Dependencies:

- [ ] https://github.com/matrix-org/matrix-spec-proposals/pull/4388
- [x] https://github.com/matrix-org/matrix-spec-proposals/pull/4341

Video demos:

**New EX scans existing EW**
![MSC4108 v2025 New EX scans existing EW](https://github.com/user-attachments/assets/51d10ce6-70a7-43b3-a881-0d1347bcf269)

**Existing EX scans new EW**
![MSC4108 v2025 Existing EX scans new EW](https://github.com/user-attachments/assets/544f70c1-ad11-4999-90db-7f0a4f64a469)

**New EX scans existing EX**
![MSC4108 v2025 New EX scans existing EX](https://github.com/user-attachments/assets/f1fdb666-92ac-400a-aaef-c8bfb5d6e17f)

---

The authors are employed by Element to write this MSC.

---

## Major revisions

The is an older version of this proposal (referred to as "2024") that has some significant differences from the current proposal.

To help avoid confusion on the status, the following is hopefully helpful:

Version name|Status|Proposal revision|Supported by|Encryption scheme|Discovery of support|Proof-of-MSC implementation status
-|-|-|-|-|-|-
2025|MSC is feature complete and being reviewed [Diff to 2024 version](https://github.com/matrix-org/matrix-spec-proposals/compare/87f8317a902cd7bc5c2d2d225f71021b3a509e2d...element-hq/oidc-qr-login)|[latest](https://github.com/matrix-org/matrix-spec-proposals/blob/element-hq/oidc-qr-login/proposals/4108-oidc-qr-login.md)|No production implementations. Experimental feature in Synapse.|HPKE|Discovery using `GET /_matrix/client/v1/rendezvous`|Fully implemented (see below)
2024|Feature complete, but will be superseded by a new version|[87f8317](https://github.com/matrix-org/matrix-spec-proposals/blob/87f8317a902cd7bc5c2d2d225f71021b3a509e2d/proposals/4108-oidc-qr-login.md)|Experimental feature in Synapse. Supported by Element Web/Desktop to generate QR. Supported by Element X to scan QR to login.|ECIES|`unstable_features`.`org.matrix.msc4108` is `true` in `/versions` response from homeserver|Fully implemented (see below)

## Proof-of-MSC implementations

### 2025 version

- Homeservers
  - [x] Synapse - experimental support for MSC4388 with `open` and `authenticated` modes: https://github.com/element-hq/synapse/issues/19433
  - [x] MAS for Device Authorization Grant from MSC4341: available since v0.8.0
- SDKs
  - [x] JS SDK - feature branch: https://github.com/matrix-org/matrix-js-sdk/pull/5219
  - [x] Rust SDK - feature branch: https://github.com/matrix-org/matrix-rust-sdk/pull/6173
- Clients
  - [x] Element X iOS & Android - automatically works when using Rust SDK from above feature branch
  - [x] Element Web - feature branch: https://github.com/element-hq/element-web/pull/32799

#### Flows implemented

Flow|Existing device|New device
-|-|-
Existing device generates QR to sign in a new device|EW or EX|EX
Existing device scans QR to sign in a new device|EX|EW

### 2024 version
Implementations for [2024 version](https://github.com/matrix-org/matrix-spec-proposals/blob/87f8317a902cd7bc5c2d2d225f71021b3a509e2d/proposals/4108-oidc-qr-login.md):

- Homeservers
	- Synapse
		- [x] Support for MSC4108 delegation to external server: https://github.com/element-hq/synapse/pull/17086
		- [x] Support for embedded MSC4108 rendezvous session: https://github.com/element-hq/synapse/pull/17056
- Clients
	- SDKs
		- [x] Rust SDK: https://github.com/matrix-org/matrix-rust-sdk/tree/poljar/qr-login
		- [x] Vodozemac: https://github.com/matrix-org/vodozemac/pull/151
		- [x] Ruma: https://github.com/ruma/ruma/pull/1800 
	- Element Web
		- [x] Support for MSC4108 on an existing device to show a QR to sign in a new device: https://github.com/matrix-org/matrix-react-sdk/pull/12370 and https://github.com/matrix-org/matrix-js-sdk/pull/4134 
	- Element X
  		- [x] Support for MSC4108 to scan a QR to sign in: https://github.com/element-hq/element-x-ios/pull/2767 and https://github.com/element-hq/element-x-android/pull/2793

![MSC4108 Element X-Web iOS demo](https://github.com/matrix-org/matrix-spec-proposals/assets/6955675/c4950806-cd41-43b9-9c5c-ecb136e6fbee)

## To-dos

The high-level to-do list for the latest version:

- [x] Update rendezvous session API to avoid issues with ETags
- [x] Update rendezvous session API so that stylistically it fits better with C-S API
- [x] Allow rendezvous session creation to require authentication
- [x] Replace [ECIES with HPKE](https://github.com/matrix-org/matrix-spec-proposals/pull/4108#pullrequestreview-3245397962) in secure channel
- [x] Revisit [QR code format](https://github.com/matrix-org/matrix-spec-proposals/pull/4108#pullrequestreview-3262435941)
- [x] Proof-of-MSC implementations for all of above
- [x] Review [thread](https://github.com/matrix-org/matrix-spec-proposals/pull/4108#discussion_r2476856870) about sequence of `protocol_accepted`
- [x] Provide a discovery mechanism that works once the proposal is stabilised

hughns committed:
Placeholder

hughns committed:
MSC4108

hughns committed:
WIP of MSC4108

hughns committed:
Auto numbers don't work on non-sequential items

hughns committed:
High level description of rendezvous protocol and consistency in payload vs message

hughns committed:
Cheat spell checker

hughns committed:
Description of rendezvous session API

hughns committed:
Add description of QR format

hughns committed:
Lint

hughns committed:
Notes on threat model

hughns committed:
Fix broken link

hughns committed:
Resolve some more TODOs

hughns committed:
Merge branch 'element-hq/oidc-qr-login' of https://github.com/matrix-org/matrix-spec-proposals into element-hq/oidc-qr-login

hughns committed:
Define POST response body explicitly

dkasak commented on proposals/4108-oidc-qr-login.md:
A better idea is probably to just include the intent in the HKDF info, making this check implicit.

Mostly commenting as a note to myself to modify this, but discussion is welcome.

poljar replied:
That would result in a worse error message. Not saying that we shouldn't remove it, just that the intent might have some usage.

hughns replied:
The reason for putting the intent in the QR is so that the Matrix client that has scanned can immediately determine if the QR code is usable for whatever the user has asked it to do.

Then, as @poljar suggests, if it is the wrong type then it can display a suitable error message.

In the QR code representation it is encoded in the existing single `mode` byte. So, removing the intent from the QR code would not reduce the size.

hughns commented on proposals/4108-oidc-qr-login.md:
I agree that the terms are not ideal. I think I chose them as they are used elsewhere in the Client-Server API specification. I'm not sure "petition" and "grant" are better though!

Another way we could describe it is by defining the disposition of the device.

So, the "initiator" could become the "un-authenticated device". And the "reciprocator" could become the "authenticated device"?

davidegirardi replied:
I think a device, the unauthenticated one, "wishes" to login and the other, authenticated, one "grants" it.

hughns replied:
> I think a device, the unauthenticated one, "wishes" to login and the other, authenticated, one "grants" it.

This "grants" terminology clashes with OIDC.

In OIDC the OIDC Provider "grants" access after the user has "consented".

Specifically, the device doesn't have the authority to grant access. Only the OIDC Provider.

davidegirardi replied:
What about  `which wishes to login, or an existing device that wishes to open for a login" then?

hughns committed:
Add Cache-Control and Pragma HTTP response headers

hughns committed:
Add error codes

hughns committed:
Formatting

hughns committed:
Whitespace

hughns committed:
More formatting

dkasak committed:
Tweaks to the QR code login crypto (#4129)

- Derive separate encryption keys and nonces for the two sides of the secure channel
- Spell out HKDF parameters in text too
- Misc style fixes

---------

Co-authored-by: Damir Jelić <poljar@termina.org.uk>
Co-authored-by: Hugh Nimmo-Smith <hughns@element.io>

hughns committed:
Add missing device id check step to sequence diagram

zecakeh commented on proposals/4108-oidc-qr-login.md:
Unless I am having trouble following this, isn't the response a JSON object with a `url`?

hughns replied:
You are quite right. This should be talking about the `url`.

Fixed in 25e8fcb2c34089efd707a98917d24a57d26d23b9.

zecakeh commented on proposals/4108-oidc-qr-login.md:
Here the location is in the JSON body, not in the `LOCATION` header, isn't it?

hughns replied:
Again, well spotted. This is fixed in e12945c5fdb9b831edc4f702862cdb433a96961b along with some other inconsistencies.

hughns committed:
Remove references to rendezvous session ID

hughns committed:
Fix POST endpoint and Location references

hughns committed:
Rendezvous sessions should have a fixed lifetime and allow enough time to complete login

hughns committed:
Set max payload size to 4KB and fix content-type as text/plain (#4134)

* Set max payload size and fix content-type as text/plain

* Set max payload size to 4KB

hughns committed:
Cross signing is mandatory

The user must having cross signing set up and the private keys available on the existing device for the process to complete.

hughns committed:
Use unstable prefix for errcode

davidegirardi commented on proposals/4108-oidc-qr-login.md:
```diff
@@ -220,7 +219,10 @@ The server should allow a minimum payload size of 10KB and enforce a maximum pay
 ###### Maximum duration of a rendezvous
 
-The rendezvous session only needs to persist for the duration of the handshake. So a timeout such as 30 seconds is adequate.
+The rendezvous session needs to persist for the duration of the login. So a timeout such as 60 seconds should be adequate.
```
I think 60 seconds is very tight if you have a slow connection, not because of bandwidth but because of latency in establishing the various connections. I'd go for 180 seconds, which I have to admit is just a guesstimate.

tonkku107 replied:
I think defining a minimum duration instead of a maximum would be preferable

davidegirardi replied:
You want a maximum duration to avoid abuse of the service.

tonkku107 replied:
Yes, but maybe the maximum should be up to the implementation while the spec defines a minimum that should work in most cases

davidegirardi replied:
I think the spec should not lean towards vulnerable implementations. A maximum duration is the way to go if you ask me.

hughns replied:
I've incorporated a max and min into https://github.com/matrix-org/matrix-spec-proposals/commit/34ade3c1ac66916430cc7e20ad3e7bb0d2518a2a

hughns committed:
The If-Match header on PUT requests contains the ETag

zecakeh commented on proposals/4108-oidc-qr-login.md:
```diff
@@ -0,0 +1,1588 @@
+HTTP response codes, and Matrix error codes:
+
+- `200 OK` - payload returned
+- `304 Not Modified` - when `If-None-Match` is supplied and the ETag does not match
```
This is actually if the ETag matches the one provided, right?

hughns committed:
Fix description of 304 GET response

poljar commented on proposals/4108-oidc-qr-login.md:
```suggestion
Sent by: new device
```

The graphs show that this is sent by the new device.

hughns replied:
Further to our conversation just now, I concur.

poljar commented on proposals/4108-oidc-qr-login.md:
`unsupported`, according to the above table, doesn't exist.

```suggestion
    "reason": "unsupported_protocol",
```

hughns committed:
Fix m.login.failure reason typo

Co-authored-by: Damir Jelić <poljar@termina.org.uk>

hughns committed:
Fix originator of m.login.declined

Co-authored-by: Damir Jelić <poljar@termina.org.uk>

hughns committed:
Use server name rather than base URL and clarify well-known discovery

dkasak commented on proposals/4108-oidc-qr-login.md:
```suggestion
CheckBytes := HKDF_SHA256(SH, "MATRIX_QR_CODE_LOGIN_CHECKCODE|" || Gp || "|" || Sp , salt=0, size=2)
```

hughns committed:
Update 4108-oidc-qr-login.md

Co-authored-by: Denis Kasak <dkasak@termina.org.uk>

Beryesa commented:
Btw can we get the inverse option too? Logging into the desktop/web client through QR on mobile, similar to Whatsapp Web or Discord log in.

hughns commented:
> Btw can we get the inverse option too? Logging into the desktop/web client through QR on mobile, similar to Whatsapp Web or Discord log in.

Yes, this proposal allows for this flow too. 👍

albertotirla commented:
this MSC allows for logging in and verifying a device with a qr code, but as it's known for some time now, qr codes present significant challenges to visually impaired people, who can't see the portion of the screen the qr code is in to scan it, because most assistive technologies aren't advanced enough for that. So then, what about visually impaired people who also want a quicker login flow?

Here's a proposal I thought of, based on what other apps do about this, like whatsapp for example:

* only the already signed in device can create a code, this removes complexity significantly and allows for more suppositions to just be encoded into the code itself
* the code is preferably numeric only, for eas of entry, and hopefully not longer than 10 digits, assuming that such is possible
* when the person logs in with such a code, it should be decomposed in more or less the same way as the qr code was, except that perhaps some options should be sent over the insecure channel, for the code to not be very long
* none of the payloads and mechanisms presented in this MSC have to change whatsoever, this can still be called qr code login, even if it's technically not really a qr code anymore

There are two more interesting alternatives I thought of, which may or may not be of more use:

* the first device generates a compressed sound signal, containing the string the qr code did, with error correction for ambient noise and whatnot, meanwhile the other device uses its microphone to record it. Afterwards, the signal is validated and potentially fixed using the error correction bands, which may have to then be stripped with an equaliser, then the rest of the communication can continue normally. If you think this is potentially discriminating people who have both a visual and a hearing impairment, my preliminary thoughts on the matter suggest that's not necesarily the case, because the other device records the signal, not the person, and if it errors due to interference, it would have errored anyway, regardless whether the person hears it
* nfc chips could also be used, if they're available on both devices. How this would work is that one would be configured as the transmiter, so to speak, it will contain the contents the qr code should have. Then, when the other nfc enabled device gets really close to the first, it gets sent that binary string, which can then be parsed according to the specification

What do you think, would any of this be possible at all? true, we can login normally, by specifying a homeserver, being redirected to the oidc provider and going through all that stuff, but just because we can't scan the qr code, we shouldn't be barred from this kind of quick login. So then, let's see if we can make matrix more accessible, from the roots rather than relying on the clients to implement something nonstandard, if at all

jotoho commented:
Some first thoughts on @albertotirla's proposals, from a random matrix user just watching the PR out of curiosity:

#### Proposal 1

This proposal might prevent QR code login from being used in the common usecase of wanting to use your logged-in smartphone/tablet to quickly login a device without a camera, such most desktop (and probably some laptop) pcs. A PC at work or school, for example.

An authentication code of format `[0-9]{10}` would only be ten billion possible combinations. Server side protections against brute-forcing could (partially) migitate risk of an attacker guessing, but my gut feeling is that that might not be enough, if the attacker has access to many IP addresses, such as via a botnet.
Especially since I don't know how large homeservers might grow in the future and an attacker might just have to guess the combination of any user currently trying to login, if they don't care about compromising a specific account.

(Also, I really hope any login is exclusively done over secured channels, such as HTTPS.)

#### Proposal 2

The sound code login idea sounds interesting, but also unsuitable as a default(-ish) login method. It might be promising as an optional alternative to QR code login.

In many scenarios intentionally causing noise is not desirable or socially acceptable.

I suspect the sounds generated might not be pleasant for humans and that many potential client devices do not have any attached microphones (or speakers). Even some smartphones intentionally do not have a built-in microphone, or have microphone access restricted.

Additionally, depending on where you are, unauthorized recording of sound-encoded authentication codes might be easier than being in a position to repliably scan a displayed QR code on a target's device.

#### Proposal 3

NFC-driven login sounds interesting, but has the restriction of likely being limited to smartphone-to-smartphone login, which cripples it's real-world usefulness.

NFC chips are only common in smartphones, to the best of my knowledge, and I currently do not see any trend to add them to many other potential matrix client devices.

albertotirla commented:
ok, so let's address this in order. I'll focus more on your issues with proposal 1, however perhaps it will be nice to modify this spec to have a pretty much plugginable architecture for how the code is generated, because it's the same information anyway

* whatsapp does it with a shorter code, it's fine, I haven't heard of anyone trying to login like that and succeeding. It also works with 2fa and authenticator apps, the procedure is similar, although arguably the user has to enter the password and such before arriving to the second factor login
* the code can be brute forced, but see the rest of this pr. I specifically said none of the other flows are modified here, so the insecure session is done the exact same way, insecure push and pull get wrapped to make secure versions, etc etc all the same
* there's a timeout for the loggin flow, like with the qr code way, but because it's a VI person most likely to use this, either the timeout is per digit which would be difficult to implement, or it should have to be enlarged to like 5 minutes or so, because the user has to both hear the digit from the phone while navigating character by character, and type it on, say, the computer
* like with the previous issues, the brute force attack will still be mitigated only at the last step, where the user is asked to check if the other side is displaying a check mark, which I presume will be read by a screenreader. If not, well, they should either cancel the login or let it expire on its own, no harm done
* qr code login wouldn't be hindered in any way by this, that will continue to work, except that now the client also shows the code below the qr one which should be scanned by the other device, see whatsapp
* when it comes to the code, even if you're logged into the phone, you can still read it char by char and type it on the computer that way. Also, as far as I know, this is possibly a flow encoded in oidc, although not like this, with secure channels and all that, so it should be supported in the future regardless
* look in the mitigations section of this pr, there's something about rate limiting there, people should implement that
The other methods aren't very important, but quickly about the sound one

* the sound wave doesn't have to be audible, or even loud for that matter, see amazon alexa vulnerabilities triggered by commands synthesized under the level of human hearing, and even so they got picked by the microphone just fine
* I haven't heard of a device which doesn't provide means of recording something, be it with a crappy microphone, in a very long time. And yes, microphones are locked down with a good reason, however the matrix client in question has to ask permission for it, it'll have to soon enough anyway, when it comes to calling
* for the someone recording the thing case, see above, rate limiting as well as that other protection about the check mark on the other device and the two digit code they have to type

dkasak commented:
A requirement of this proposal is that the QR code needs to contain 1) a 32-byte Curve25519 public key 2) the URL of the rendezvous server session. This makes the proposal to turn it into a short numeric code a non-starter.

(Also note that the standard way of commenting on MSCs is by using the [`Files changed` view](https://github.com/matrix-org/matrix-spec-proposals/pull/4108/files) and leaving comments on particular lines, by clicking and dragging on the blue + sign that appears when you hover over a line. However, given the context of accommodating visually impaired people, this may pose challenges, so consider this as just an FYI.)

albertotirla commented:
for the first issue, it could be relatively simple to fix

* the binary string is first compressed, with lzma or another high compression ratio algorythm. For reproduceability, perhaps pregenerate a compression index or dictionary for the known parts of the string and store it in the MSC, for clients to directly use that one, so that some space in the compressed byte stream is saved because the header isn't required anymore
* just like we can represent a cryptographic key as a set of emojis, I'm sure we can represent the compressed data stream as a set of numbers. May be longer than 10 digits, but considering the code and the compression algorythm involved, it couldn't be much longer either
* or perhaps the sound based carrier wave is indeed a slightly better idea, with some alterations, as discussed in the previous comment
* maybe the code could be shortened by having the user type their homeserver before logging in, however that might not be better than normal login
* what if we require devices to connect to each other via bluetooth, or find one another via the network, zeroconf discover stile and send the code that way?

Anyone any other ideas though? I'm trying to come up with a scheme for us to also be able to use this kind of login, but surely I might be missing some idea which is better, after all whatsapp used a similar method and it works for them, perhaps we should simplify our flows?

For the second, I thought that might be the case, and github isn't accessible enough to be able to comment on multiple lines, just one at a time. However, when I asked that in a matrix room, I'll leave the name of that room out to not incriminate people who don't want to be incriminated, people told me to do it this way, because I didn't know on what line to put the comment. Either way, I'll know for the future, in case there's an actual line to comment on

albertotirla commented:
so, given all the feedback in the previous comments, and also due to a friend of mine who gave me an idea, I think I have a better grasp of what we would need to make this work

* the binary string is generated, as usual, nothing to see here
* the blob goes through a base64 encoder
* it is converted into sound, like I said before. We did establish that just using sound as a carrier wave for raw data could be very unpleasent for the human ear, and upon further discussions with the before mentioned friend, I came to realise that if we do it above the level of human hearing and blast the data that way, we could trip some emergency alert tones and stuff like that, so that's out of the question as well. We should use a nevertheless standardised method for generating those tones, something which doesn't hurt the human ear and something which can't easily be confounded for other kind of traffic over sound waves. In this instance, I recommend morse code, because
  * it's pretty ubiquitous, it's well understood
  * we can use a more specific frequency, so that we can better isolate ambient noise and whatnot, we don't have to push that data over all bands with error correction bands as well, because now the sound has a well-defined shape, like the qr code for that matter
  * it can encode any length of data, as long as the user can wait for long enough for the thing to be played, which is a small concern, but since morse code can be pretty fast, I don't think we should wory
  * it doesn't require simplification of the protocol, because the exact same binary string of the qr code, with some transformations perhaps, can be encoded in morse no problem
* any device can generate it, and likewise any device can scan it, as long as it has at least one microphone, obviously the more microphones, the more likely that the signal is recieved without distortions and other channels from the recording can technically be used as error correction in that case
* once the binary string gets on the other side, the mechanisms are executing as planned in the rest of this spec, this problem is essentially ways in which to get that binary string from one device to another, one solution to that being a qr code, the other is this
* after successful login, because this is recording sound, perhaps have the code that has to be typed be three digits long instead of two, to mitigate some of the issues above

If this is to be deemed acceptable for most people here, I recommend we make it so that we can't find ourselves in the situation where one client only implements morse code, another only qr and then it's users who suffer. So, I think clients who implement the qr flow should be required by the spec to also implement the morse code for accessibility reasons, but those who implement only morse code don't have to also implement qr code, because anyone can still use the morse code login flow even if they can see to scan a qr code.

Is this a better idea? perhaps it's time we try something out and see if it could work, are there issues I overlooked? or maybe, who knows, is it worse than those which came before it?

jotoho commented:
@albertotirla 

The spec isn't law. Even if the spec said implementations supporting QR code also need to support audio code login, I think neither the spec nor servers reasonably can (or should) enforce that rule.

And if I understand correctly, one of the goals of moving auth away into MAS and using OIDC is to make implementing compliant clients easier.
Adding a hard requirement to support a login method that at least some maintainers might not be interested in developing/maintaining, seems counterproductive to that goal.

So if the audio login method is going to, at most, be an (optional) alternative to QR code login anyways, I feel like discussions on that probably shouldn't derail discussion/progress on this QR code login MSC.
If there's any interest from implementation maintainers it might be better for you to propose a seperate MSC for that. If implementations aren't interested, it's dead in the water anyways.

I've said my two cents, so I'll stay out of the discussion from here on. :slightly_smiling_face:

albertotirla commented:
yeah, people told me throughout today to make a new MSC for this, which is what I will do soon enough, not sure on specific details unfortunately. So yeah, for the other people beside me who want quick login to be accessible to everyone, and especially if you think the morse code idea has some merit to it, stay tuned!

erikjohnston commented on proposals/4108-oidc-qr-login.md:
I think its worth noting that QR login allows an attacker with temporary access to an unlocked client to extend that to ongoing access, i.e. if you leave your phone unlocked briefly someone could quickly use QR code login to login on their device.

It also makes it easier to get the private keys of the user from an unlocked client, as you can login with a new device, extract the keys from that, and logout again to cover your tracks.

Sophisticated attackers can today use specialist equipment to extract private keys and access tokens from the memory of a process. However, a) that is a much higher bar for attack, and b) cloning an access token will quickly be detected via refresh tokens. 

I think we should add some recommendations to help mitigate this, something like:
- Servers SHOULD send new device login notifications to other devices or email
- Native clients SHOULD gate QR code login behind some form of extra protection, e.g. biometrics on mobile apps. These should be minimally invasive though as otherwise it heavily erodes the benefit of using QR code login in the first place. We don't necessarily think this protection is worth while on web clients, as its trivial to access the devtools to extract the secrets directly and/or bypass any extra protections.

tonkku107 commented on proposals/4108-oidc-qr-login.md:
To my knowledge, none of the clients listed in the PR description support this new device mode. Should there be a client that implements this before merging to the spec?
(I commented on the wrong mode at first, this is confusing)

uhoreg replied:
If there are no available implementation of `0x03`, then that seems like a significant omission, and it may block this MSC getting FCP'ed

noelportillo reviewed (commented):
Good job

hughns committed:
Update to match spec 1.15 and MSC4341

hughns committed:
Update links from spec 1.10 to 1.15

hughns committed:
Feedback from review

See https://github.com/matrix-org/matrix-spec-proposals/pull/4108#discussion_r1954778980

hughns committed:
Min and mix rendezvous timeouts

See https://github.com/matrix-org/matrix-spec-proposals/pull/4108#discussion_r1575855901

turt2live commented on proposals/4108-oidc-qr-login.md:
```diff
@@ -0,0 +1,1593 @@
+which the two devices can exchange the necessary data. This session is described as "insecure" as it provides no
+end-to-end confidentiality nor authenticity by itself---these are layered on top of it.
+
+#### High-level description
```
Editorial: Having a sequence diagram early in the proposal here would help a lot in understanding the flow.

turt2live commented on proposals/4108-oidc-qr-login.md:
```diff
@@ -0,0 +1,1593 @@
+
+HTTP request body:
+
+- any data up to maximum size allowed by the server
```
we should specify what this limit is rather than leaving it up to the implementation, otherwise clients will have to guess & check, or assume.

Likewise on other HTTP endpoints/methods.

turt2live commented on proposals/4108-oidc-qr-login.md:
````diff
@@ -0,0 +1,1593 @@
+}
+```
+
+##### Send a payload to the rendezvous session: `PUT <rendezvous session URL>`
````
I'm very worried about unauthenticated endpoints accepting data the server needs to persist indefinitely. Rate limits, size limits, and longevity limits need to be described by the MSC.

turt2live commented on proposals/4108-oidc-qr-login.md:
This is a compelling alternative, honestly. It causes the caller to be authenticated, *and* accept the Terms and Conditions explicitly rather than implicitly (which is questionable as-is). It also means traditional safety tooling can be applied to the user. 

I think it's also fair to say that QR code login isn't available on servers which disable guest access. Single-user (or low-user) homeservers can enable it as-needed, and private homeservers can apply network conditions or registration requirements to disable guest access to the wider internet. Public homeservers are likely to already want/have guest access enabled.

Being able to "upgrade" a guest account to an existing account would be ideal as well, though for purposes of scoping can/should be deferred to yet another future MSC.

erikjohnston replied:
One thing that I think is not immediately obvious from this MSC is that there are two ways of starting a QR code session:
1. Generating a QR code on a logged in device. This device can create a rendezvous channel on its local HS, and so this can be authenticated. 
2. Generating a QR code on a fresh device. The client has no idea what account or HS its going to sign it to, as its the other device that is logged in (that will scan the QR code). The client then creates a rendezvous channel on a static pre-configured server (e.g. matrix.org), but with no authentication.

This has some ramifications:
1.  We can default to homeservers require authentication to *create* rendezvous channels, which helps T&S for the majority of HSes. 
2. Clients that want to generate QR codes when logged out must pre-configure a homeserver to accept unauthenticated rendezvous channels. This is where we need to be careful with T&S, but at least we've limited the scope and can potentially limit it further (maybe make it so that the rendezvous channel redirects to the actual HS or something)
3. I don't think that using to-device channels really help in this situation. To handle the case of non-logged in devices you'd need to allow automatic registration of a temporary user and device on a 3rd party HS (and we would strongly want to avoid having to manually accept terms for a 3rd party server from a product PoV), which just feels way more heavyweight than allowing locked down rendezvous channels.


I'm going to add a new comment about adding auth

turt2live commented on proposals/4108-oidc-qr-login.md:
```diff
@@ -0,0 +1,1593 @@
+- `201 Created` - rendezvous session created
+- `400 Bad Request` (``M_MISSING_PARAM``) - either `Content-Length` and/or `Content-Type` was not provided.
+- `400 Bad Request` (`M_INVALID_PARAM`) - an invalid `Content-Type` was given.
+- `403 Forbidden` (``M_FORBIDDEN``) - forbidden by server policy
```
is 403 how a server can opt-out of this? Because the Matrix specification requires servers to implement all endpoints, without an opt-out mechanism it's possible for *any* server to become abused for this channel, even if the two users talking have nothing to do with that server.

Being able to apply some restrictions, like only allowing logged-in users to call this endpoint, is essential.

turt2live reviewed (changes requested):
Overall I think I prefer the difference described by https://github.com/matrix-org/matrix-spec-proposals/commit/456beb61b0fd31f3992252f19d41be88732dd35d#r165405539 

Though, the MSC's premise is scary: open, unauthenticated, writes to a server is something we avoid in our protocol. Limits will help remove a lot of the scary factor, but I'm not sure it will be enough. I strongly encourage exploring the to-device alternative as an avenue for this feature.

I also find this MSC hard to review due to its size, so have paused my review around line 250 or so. I'd suggest breaking the MSC down to at least core transport (channel + security), then opening dedicated MSCs for the application traffic/uses. 

The github UI also doesn't appear to like this MSC too much, possibly due to the heavy usage of mermaid charts - ASCII versions may be needed?

erikjohnston commented on proposals/4108-oidc-qr-login.md:
```diff
@@ -0,0 +1,1593 @@
+##### Authentication
+
+These API endpoints do not require authentication because trust is established at the secure channel layer which is
+described later.
```
One thought for restricting what you can do with unauthenticated rendezvous channel is to change it to instead work something like:

1. Unauthenticated device (UD) creates a "redirect channel" on HS1 and sets that in the QR code.
2. The authenticated device (AD) creates a rendezvous channel on HS2.
3. HS2 POSTS to the redirect channel on HS1 with the homeserver and rendezvous channel ID. HS1 validates its from HS2.
4. HS1 returns the homeserver (HS2) and rendezvous channel ID to UD, who then uses that channel as normal.

This has the following properties:
1. It limits how much information can be persisted on an unauthenticated channel. We can severely restrict the size of the request ID for example.
2. An abuser must use a domain they own if they want to encode dodgy data in the rendezvous channel ID. We can then ban abusive domains.
3. An unauthenticated device can only receive information, rather than create a 2-way channel. Not sure that's at all useful thing to assert, but it is nonetheless a property.

I think this sort of flow would reduce potential abuse vectors, but equally makes things more complicated and may not be worth it.

hughns committed:
Add table of contents

uhoreg commented on proposals/4108-oidc-qr-login.md:
I don't think any of this is specific to QR codes, and could be used by any other authentication mechanism that wants to use the rendezvous service similarly in the future.  (NFC or something?)  So I would suggest dropping the `QR_CODE` from all the strings.

hughns replied:
It was a deliberate choice to include the `QR_CODE` in the prefix so that the protocol is bound to the specific application (the QR code login).

Whilst the could be some future usage of the secure channel scheme (e.g NFC based login) this would be speculation at this point.

I will add a clarification about this intention to the MSC.

hughns replied:
Clarified in ad719369197fda5459d55759949932503f25f529

uhoreg commented on proposals/4108-oidc-qr-login.md:
```diff
@@ -0,0 +1,1593 @@
+
+- common headers as defined above
+
+##### Receive a payload from the rendezvous session: `GET <rendezvous session URL>`
```
Could the rendezvous server just require some header, maybe an `Authorization` header even if it uses a dummy value, to prevent someone from accidentally opening the URL?

dkasak replied:
I like uhoreg's proposal. It feels much nicer than collapsing everything into a single SOAP-y endpoint.

hughns replied:
Some thoughts:

- Compared to the original proposal (which allowed any content type) this current one is better in that it restricts to `application/json` and only the `data` string is open to abuse. However, a risk does reamin.
- Even with a `POST` it is simple enough to have a form hosted somewhere that a victim's browser could navigate to.
- Could we instead treat this as a browser content/security policy issue> For example we could have the server check the [Sec-Fetch-Mode](https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Headers/Sec-Fetch-Mode) header and return some error if the mode is `navigate`. I am thinking of this akin to how CORS works where it is a reasonable starting point to assume that a victim has a well-behaved browser that is setting headers as expected.

Thoughts?

hughns committed:
Suggestions from @uhoreg

hughns committed:
Fix incorrect string

hughns committed:
All nonces start at 0

hughns commented on proposals/4108-oidc-qr-login.md:
In step 4 above it says:

> - If the user consents in the next step then the new device will receive an `access_token` and `refresh_token` etc. as
normal as per [MSC4341].

The is also a note above which says "as per RFC 8628 / MSC4341".

hughns committed:
Make rendezvous API more like rests of Client-Server API

hughns committed:
Make 429 errcode be M_LIMIT_EXCEEDED

dkasak commented on proposals/4108-oidc-qr-login.md:
```diff
@@ -0,0 +1,1609 @@
+even arbitrary network participants which possess the rendezvous session ID and server name.
+To provide a secure channel on top of this insecure rendezvous session transport, we propose the following scheme.
+
+This scheme is essentially [ECIES](https://en.wikipedia.org/wiki/Integrated_Encryption_Scheme#Formal_description_of_ECIES)
+instantiated with X25519, HKDF-SHA256 for the KDF and ChaCha20-Poly1305 (as specified by
+[RFC8439](https://datatracker.ietf.org/doc/html/rfc8439#section-2.8)) for the authenticated encryption. Therefore,
+existing security analyses of ECIES are applicable in this setting too. Nevertheless we include below a short
+description of our instantiation of ECIES and discuss some potential pitfalls and attacks.
```
If we were designing this today, I'd use [RFC 9180 HPKE](https://datatracker.ietf.org/doc/rfc9180/) instead. It's conceptually the same construct as ECIES but better specified, backed by the CFRG/IRTF and used in MLS.

hughns committed:
Add note bout message prefix

hughns committed:
Remove further references to rendezvous server

hughns committed:
Fix more references to OAuth and MSC4341

hughns committed:
More consistency on SecureSend/SecureReceive

hughns committed:
Split out steps and reorder for clarity

hughns committed:
And example for `device_already_exists` outcome

hughns committed:
Rendezvous authentication and optionality

Make rendezvous API optional and return 404.
Make authentication on creating rendezvous optional and return 403.
Add client header filtering for unsafe content.
Clean up for readability.

hughns committed:
Add alternative about unauthenticated device creating "redirect channel"

hughns commented on proposals/4108-oidc-qr-login.md:
```diff
@@ -0,0 +1,1593 @@
+##### Authentication
+
+These API endpoints do not require authentication because trust is established at the secure channel layer which is
+described later.
```
I have done this in 9dab408720d89ad4d96dabc7fb3ba353f8072aea

hughns committed:
Fix description of discovery steps

hughns committed:
Move QR format part of proposal to where it sits in the flow

hughns committed:
QR code clarifications

Clarify original thinking behind QR format.
Add notes about versioning and unstable prefixes

hughns committed:
Wording  on new/existing device

hughns committed:
Remove another reference to reciprocate

hughns committed:
Reinstate note about long poll for future

dkasak commented on proposals/4108-oidc-qr-login.md:
```diff
@@ -0,0 +1,1730 @@
+The QR codes to be displayed and scanned using this format will encode binary strings in the general form:
+
+- the ASCII string `MATRIX`
+- one byte indicating the QR code version (must be `0x02`)
```
What is the meaning and purpose of this field? What does the value 0x02 represent, i.e. what practical insight do you gain upon reading the value 0x02 in a message?

dkasak commented on proposals/4108-oidc-qr-login.md:
```diff
@@ -0,0 +1,1730 @@
+  - two bytes in network byte order (big-endian) indicating the length in bytes of the server name as a UTF-8 string
+  - the server name as a UTF-8 string
+
+If a new version of this QR sign in capability is needed in future (perhaps with updated secure channel protocol) then
+additional "mode" values can be added.
```
This feels like a hack. The modes were meant to encode directionality of the two participants, so it IMO makes little sense to add additional ones.

It can certainly be made to work for this particular case, but I'm beginning to wonder why this MSC does not have a single protocol versioning field. In light of https://github.com/matrix-org/matrix-spec-proposals/pull/4108/files#r2375468437, I believe the QR code version field should be clarified to be this missing protocol version field for the protocol described by this MSC.

hughns committed:
Revert removal of public key from example QRs

hughns committed:
Clarify 4KB

hughns committed:
Clarifications around public key in QR

hughns committed:
GitHub doesn't like the link syntax I used

Johennes commented on proposals/4108-oidc-qr-login.md:
When implementing this in matrix-rust-sdk, @poljar raised the question of whether the order of opening the browser on the existing device and sending the `m.login.protocol_accepted` message to the new device matters. Should we first send the acknowledgment so that the new device can start polling and only then open the browser to let the user consent?

I _think_ it doesn't make a huge difference but wanted to double check it here.

(Full discussion in https://github.com/matrix-org/matrix-rust-sdk/pull/5801#discussion_r2469444276)

hughns replied:
I have some recollection on this topic: 

On receipt of the verification URI the existing device might not be able to open it. Perhaps the operating system denies the request, or it turns out that the isn't a system browser available.

As such I *think* the intention was that `m.login.protocol_accepted` is only sent after the device knows that it can proceed with showing the URI.

But the MSC doesn't say this, and nor does the seem to be a suitable `m.login.failure` reason code defined.

If this was defined then I don't think it matters if the new device starts polling before it has received the `m.login.protocol_accepted`. But, it would need to stop and do something sensible if it received a `m.login.failure` instead of `m.login.protocol_accepted`.

hughns committed:
Update unstable prefixes including on QR code

hughns committed:
Update QR code format under "type" 0x03

The byte immediately following the `MATRIX` prefix is repurposed as a type which gives a more sensible way to namespace in future.

This also means that we can call the second byte the "intent" very clearly rather than having to call it "mode" to match the existing cross verification QR spec.

hughns commented on proposals/4108-oidc-qr-login.md:
```diff
@@ -0,0 +1,1751 @@
+  - two bytes in network byte order (big-endian) indicating the length in bytes of the rendezvous session ID as a UTF-8
+  string
+  - the rendezvous session ID as a UTF-8 string
+- the [server name](https://spec.matrix.org/v1.15/appendices/#server-name) of the homeserver encoded as:
```
Reflecting on this some more: as discovery of the OAuth issuer is now done via the `/auth_metadata` [endpoint](https://spec.matrix.org/v1.16/client-server-api/#get_matrixclientv1auth_metadata) on the base URL (as opposed to from .well-known/matrix/client on the server name like it was when this proposal was originally written) I think it definitely makes sense to use the base URL. I will update the proposal to do this.

hughns committed:
Use base URL in QR code and m.login.protocols message

hughns committed:
Update QR examples to match description

hughns committed:
Split secure channel into MSC4388 + add intro diagram

hughns committed:
Update proposals/4108-oidc-qr-login.md

Co-authored-by: networkException <github@nwex.de>

hughns committed:
Remove old QR codes as now part of MSC4388

hughns committed:
Add unable_to_open_verification_uri failure reason

hughns committed:
Add an example of where user_cancelled can be used

hughns committed:
Complete the "potential issues" section

turt2live reviewed (approved):
I've done a brief requirements review, and this proposal appears to be ready for FCP with a dependency on MSC4388 (other dependencies are already merged)

turt2live commented on proposals/4108-oidc-qr-login.md:
@mscbot concern Dependency on MSC4388 outstanding

richvdh commented on proposals/4108-oidc-qr-login.md:
```diff
@@ -0,0 +1,904 @@
+
+Otherwise the new device waits to be informed by receiving an `m.login.protocols` message from the existing device.
+
+The existing device would need to determine which "protocols" are available for the new device to use.
```
```suggestion
The existing device would need to determine which "login protocols" are available for the new device to use.
```

maybe?

richvdh commented on proposals/4108-oidc-qr-login.md:
```diff
@@ -0,0 +1,904 @@
+
+We use the `SecureSend` and `SecureReceive` operations from [MSC4388] which are sent via the out-of-band channel.
+
+1. **Homeserver discovery**
```
These get rendered as single-element `<ol>` lists, which I'm finding surprisingly visually distracting.

Assuming we don't want to make the whole proposal into an actual list, maybe we should just reach for the fourth level of heading?

```suggestion
#### 1. Homeserver discovery
```

hughns replied:
Done in 754828108de636ffc23ef6e8a56ff8ede3c81159

richvdh commented on proposals/4108-oidc-qr-login.md:
```diff
@@ -0,0 +1,904 @@
+The existing device would need to determine which "protocols" are available for the new device to use.
+
+Currently this could only be `device_authorization_grant` meaning the homeserver supports the
+`urn:ietf:params:oauth:grant-type:device_code` grant type.
```
A general comment on the text of the MSC: I don't find these single-sentence paragraphs particularly easy to read. Can we maybe combine related ideas into paragraphs?

hughns replied:
I have attempted to strike a better balance in eeff6bb864b133b813454d41b52f44a9929fd519

Please say if you think even more is needed.

richvdh commented on proposals/4108-oidc-qr-login.md:
```diff
@@ -0,0 +1,904 @@
+Currently this could only be `device_authorization_grant` meaning the homeserver supports the
+`urn:ietf:params:oauth:grant-type:device_code` grant type.
+
+If it is available then the existing device informs the new device by sending the `m.login.protocols` message with the
```
Is "it" the device_code grant type?

```suggestion
If that grant type is available, then the existing device informs the new device by sending the `m.login.protocols` message with the
```

richvdh replied:
What happens if that grant type is *not* available? Does the existing device just not send the `m.login.protocols` message?

richvdh replied:
> What happens if that grant type is _not_ available? Does the existing device just not send the `m.login.protocols` message?

Maybe the existing device should just not offer to show/scan the QR code? Would be good to be explicit.

hughns replied:
Addressed in e04dbc9626eb19601a4cd601395064d6a4503dba

richvdh commented on proposals/4108-oidc-qr-login.md:
```diff
@@ -0,0 +1,904 @@
+
+2. **New device checks if it can use an available protocol**
+
+The existing device then undertakes steps to determine if it is able to work with the homeserver.
```
```suggestion
The new device then undertakes steps to determine if it is able to work with the homeserver.
```

I think?

richvdh commented on proposals/4108-oidc-qr-login.md:
````diff
@@ -0,0 +1,904 @@
+    ...
+    "device_authorization_endpoint": "https://auth-oidc.lab.element.dev/oauth2/device"
+}
+```
````
please indent all this by two spaces to avoid interrupting the bullet list.

Likewise for the other bullets below.

hughns replied:
Done in b6dc96c593979370e7d93a9d6fe67c5fcc5b9fc5

richvdh commented on proposals/4108-oidc-qr-login.md:
```diff
@@ -0,0 +1,904 @@
+
+- parses the [Device Authorization Response](https://datatracker.ietf.org/doc/html/rfc8628#section-3.2) above
+
+At this point the new device knows that, subject to the user consenting, it should be able to complete the login
```
```suggestion
At this point the new device knows that, subject to the user consenting, it should be able to complete the login.
```

richvdh commented on proposals/4108-oidc-qr-login.md:
```diff
@@ -0,0 +1,904 @@
+- In accordance with [RFC8628](https://datatracker.ietf.org/doc/html/rfc8628#section-3.3.1) the new device must display
+the `user_code` in order that the user can confirm it on the homeserver if required.
+- The new device then starts to poll the homeserver by making
+[Device Access Token Requests](https://datatracker.ietf.org/doc/html/rfc8628#section-3.4) using the interval and bounded
```
which interval?

hughns replied:
Clarified in b6dc96c593979370e7d93a9d6fe67c5fcc5b9fc5. It's from the Device Authorization Response.

richvdh commented on proposals/4108-oidc-qr-login.md:
```diff
@@ -0,0 +1,904 @@
+[Device Access Token Requests](https://datatracker.ietf.org/doc/html/rfc8628#section-3.4) using the interval and bounded
+by `expires_in`.
+
+The above is as per [MSC4341].
```
Now that MSC4341 is canon, it might be nice to point to https://spec.matrix.org/v1.18/client-server-api/#device-authorization-flow instead.

richvdh replied:
Likewise other mentions of MSC4341

hughns replied:
Done in 76a9baec8bb2cf1f0eac030ce6852d4f421c7d4a

richvdh commented on proposals/4108-oidc-qr-login.md:
```diff
@@ -0,0 +1,904 @@
+            rect rgb(191, 223, 255)
+note over N,E: This step is duplicated from the previous section for readability
+              N-->>+E: SecureSendReceive({ "type": "m.login.success" })
+            end
```
Alanis Morissette would like a word

<img width="1236" height="252" alt="Image" src="https://github.com/user-attachments/assets/1afc5cf1-a3ea-4f8f-8e14-1ed163d02b21" />

hughns replied:
To be fair, this how it renders for me at the moment:

<img width="622" height="144" alt="Image" src="https://github.com/user-attachments/assets/c72cf7ce-fb17-4192-9272-a1462975a20b" />

More generally the mermaid rendering is very unreadable in dark mode. I'm not sure why it isn't more sensible.

hughns replied:
I've removed the iron note in 09d03a8f17e0a4bcde2ec0d0d58da95e66cb32ce as I don't think it is actually needed.

I've checked how the diagrams are rendered depending on GitHub theme. What has caught me out is that there is a [bug](https://github.com/orgs/community/discussions/194348) in the mermaid rendering where it doesn't respect the system theme.

If you have the GitHub theme explicitly set to dark then it renders just fine. Light more is always fine.

richvdh commented on proposals/4108-oidc-qr-login.md:
```diff
@@ -0,0 +1,904 @@
+    participant HS as Homeserver
+
+
+    rect rgba(255,0,0, 0.1)
```
is there a significance to the colour coding? Do the colours carry over between sequence diagrams?

hughns replied:
Yes, the idea was to highlight the difference between the two variants.

I've added a note to explain this in baf8b3f0d5319a0fb812e79270e61ce265dff9fa.

In 09d03a8f17e0a4bcde2ec0d0d58da95e66cb32ce I have also removed the background from the other diagrams where it adds nothing.

richvdh commented on proposals/4108-oidc-qr-login.md:
```diff
@@ -0,0 +1,904 @@
+    },
+    "backup": {
+        "algorithm": "foobar",
+        "key": "$base64_of_the_backup_recovery_key",
```
We don't use the term "recovery" in spec (https://spec.matrix.org/v1.18/client-server-api/#decryption-key)

```suggestion
        "key": "$base64_of_the_backup_decryption_key",
```

richvdh commented on proposals/4108-oidc-qr-login.md:
```diff
@@ -0,0 +1,904 @@
+
+3. **New device cross-signs itself and uploads device keys**
+
+On receipt of the `m.login.secrets` message the new device can store the secrets locally
```
```suggestion
On receipt of the `m.login.secrets` message the new device can store the secrets locally.
```

richvdh commented on proposals/4108-oidc-qr-login.md:
```diff
@@ -0,0 +1,904 @@
+It can then use a single request to upload the device keys and cross signing signature. This removes the chance of other
+devices seeing the new device as unverified, incorrectly prompting the user to verify the already verified device.
+
+The request would look just like any other `/keys/upload` request, it would just include one additional signature, the
```
```suggestion
The request would look just like any other `/keys/upload` request; it would just include one additional signature, the
```

richvdh commented on proposals/4108-oidc-qr-login.md:
```diff
@@ -0,0 +1,904 @@
+already, or the homeserver must support and allow dynamic client registration as described in the [spec](https://spec.matrix.org/v1.15/client-server-api/#client-registration).
+
+The feature is also only available where a user has cross-signing set up and the existing device to be used has the
+Master Signing Key, Self Signing Key and User Signing Key stored locally so that they can be shared with the new device.
```
Is this strictly correct? it seems like the login part would still work?

hughns replied:
The login part *could* work, but it has been deliberately excluded from the scope of this MSC.

I have attempted to clarify by adding an alternative in f9106ceb943c7ef387dd20c1fbfb9f484e51a9f9

richvdh commented on proposals/4108-oidc-qr-login.md:
```diff
@@ -0,0 +1,904 @@
+Against:
+
+- The existing device needs to wait for the new device to upload the device keys for it to sign the new device.
+- Takes several round-trips for the secrets to be be shared which will add latency to the overall flow
```
```suggestion
- Takes several round-trips for the secrets to be be shared which will add latency to the overall flow.
```

richvdh commented on proposals/4108-oidc-qr-login.md:
```diff
@@ -0,0 +1,904 @@
+- The existing device needs to wait for the new device to upload the device keys for it to sign the new device.
+- Takes several round-trips for the secrets to be be shared which will add latency to the overall flow
+- The backup cannot be immediately enabled since we received the backup version as well, something the `m.secret.send`
+mechanism does not offer.
```
Is this trying to say that we need to wait for a `GET /room_keys` request to the homeserver?

```suggestion
- Key backup upload cannot be enabled until we make a `GET /room_keys` request to the homeserver, in order to receive the receive the key backup version.
```

TBH I'm not sure that's a bad thing: it gives us a chance to ensure that the decryption key we were sent matches the public key for the backup version.

(Also, if all we want is the backup version, we could fix that in secret-sharing too, so I don't think this is an effective argument against using secret sharing for QR code login)

Possibly I'm missing something?

poljar replied:
Yes, I think it's trying to say this.

> TBH I'm not sure that's a bad thing: it gives us a chance to ensure that the decryption key we were sent matches the public key for the backup version.

I would argue that due to the somewhat synchronous nature of QR code login, meaning you are right now talking to your own trusted device, you're not required to do this check in this instance.

> (Also, if all we want is the backup version, we could fix that in secret-sharing too, so I don't think this is an effective argument against using secret sharing for QR code login)

The backup version wouldn't be as useful in the secret-sharing scenario. Secret-sharing travels over to-device messaging which might not arrive immediately. As such double checking that the version is correct becomes a necessity.

That being said, I think there are enough other arguments in this section. I think we can just remove this one if you remain unconvinced of this argument.

richvdh replied:
Resolving in favour of a [new thread](https://github.com/matrix-org/matrix-spec-proposals/pull/4108#discussion_r3504594875)

richvdh commented on proposals/4108-oidc-qr-login.md:
```diff
@@ -0,0 +1,904 @@
+1. Check if the homeserver has a rendezvous session API available by attempting a POST to the create rendezvous endpoint
+   from [MSC4388].
+
+For a new device it would need to know the homeserver ahead of time in order to do these checks.
```
What does this mean in practice? AIUI, the new device *doesn't* know the homeserver ahead of time, so can't do the checks?

hughns replied:
I think I have clarified this in 7239d2b7d259fac24405e5da2474b81e341150c0

richvdh commented on proposals/4108-oidc-qr-login.md:
```diff
@@ -0,0 +1,904 @@
+already, or the homeserver must support and allow dynamic client registration as described in the [spec](https://spec.matrix.org/v1.15/client-server-api/#client-registration).
+
+The feature is also only available where a user has cross-signing set up and the existing device to be used has the
+Master Signing Key, Self Signing Key and User Signing Key stored locally so that they can be shared with the new device.
```
Shouldn't this whole thing form part of the main proposal, before "Login via OAuth 2.0 Device Authorization Grant from MSC4341"?

hughns replied:
Maybe... I've moved it up in 7239d2b7d259fac24405e5da2474b81e341150c0 and expanded it to more explicitly cover the different scenarios.

richvdh reviewed (approved):
Generally this seems like an excellent, well-written proposal. I particularly appreciated the separate "Message reference" section as a way to separate the protocol object definitions from the description of the flow.

hughns committed:
Apply suggestions from code review

Co-authored-by: Richard van der Hoff <1389908+richvdh@users.noreply.github.com>

hughns committed:
Use more header styles

hughns committed:
Merge branch 'element-hq/oidc-qr-login' of https://github.com/matrix-org/matrix-spec-proposals into element-hq/oidc-qr-login

hughns committed:
Clarify checking of grant type availability

hughns committed:
Fix indent in bullet list + clarify fields coming from Device Authorization Response

hughns committed:
Combine related sentences into paragraphs

hughns committed:
MSC4341 has been merged an device authorization flow is in the spec

hughns committed:
Some wording clarifications

hughns committed:
Move discovery to main part of proposal and expand

hughns committed:
Add alternative about not requiring cross signing

hughns committed:
Remove background from sequence diagrams where colour has no meaning

hughns committed:
Explain colour coding in the sequence diagrams

richvdh commented on proposals/4108-oidc-qr-login.md:
```diff
@@ -172,7 +172,7 @@ The steps are as follows:
 
 - sends a [RFC8628 Device Authorization Request](https://datatracker.ietf.org/doc/html/rfc8628#section-3.1) to the homeserver
-  using the `device_authorization_endpoint` as described by [MSC4341]:
+  using the `device_authorization_endpoint` as described by the [device authorization grant]:
```
I think this was intended to be a link?

```suggestion
  using the `device_authorization_endpoint` as described by the [device authorization flow]:
```

richvdh commented on proposals/4108-oidc-qr-login.md:
````diff
@@ -0,0 +1,930 @@
+```
+
+The sequence for steps 1 to 3 is as follows: (the sequence depending on which device  has scanned the code varies for
+readability. The steps that are the same have a green background, the steps that are different have a red background)
````
> the sequence depending on which device  has scanned the code varies for readability.

I am struggling to parse this sentence.

hughns replied:
In d16d06cca21eea89b7fb05ff3d5f149ea53423ce I have reworded so that it hopefully actually makes sense now.

richvdh commented on proposals/4108-oidc-qr-login.md:
````diff
@@ -0,0 +1,930 @@
+
+```mermaid
+sequenceDiagram
+    title: Variant: New device scanned QR code
````
This title repeats the heading just above the diagram, so I don't think it adds any clarity. Suggest getting rid of one of the titles.

hughns replied:
Done in a0488e5a9c6d4aea7ba372c3cfef61b3e5ac0e8e

richvdh commented on proposals/4108-oidc-qr-login.md:
````diff
@@ -0,0 +1,930 @@
+
+```mermaid
+sequenceDiagram
+    title: Variant: Existing device scanned QR code
````
likewise, repetition.

hughns replied:
Done in a0488e5a9c6d4aea7ba372c3cfef61b3e5ac0e8e

richvdh reviewed (approved):
LGTM other than some minor points of clarification.

richvdh commented on proposals/4108-oidc-qr-login.md:
```diff
@@ -0,0 +1,930 @@
+- The existing device needs to wait for the new device to upload the device keys for it to sign the new device.
+- Takes several round-trips for the secrets to be be shared which will add latency to the overall flow.
+- The backup cannot be immediately enabled since we received the backup version as well, something the `m.secret.send`
+  mechanism does not offer.
```
This is still somewhat confusingly worded. (See [previous thread](https://github.com/matrix-org/matrix-spec-proposals/pull/4108#discussion_r3305677813))

Following @poljar's comments on that thread, maybe we can just note that changes *could* be made to the `m.secret.send` mechanism if this was the only problem:

```suggestion
- Key backup upload cannot be enabled until we make a `GET /room_keys` request to the homeserver, in order to receive the receive the key backup version (though the `m.secret.send` mechanism could potentially be modified to provide this information).
```

hughns committed:
Add threat modelling section

hughns committed:
Add "Shoulder-surfing to sign in to attacker account" security consideration

hughns committed:
Tidy

hughns committed:
Update proposals/4108-oidc-qr-login.md

Co-authored-by: Richard van der Hoff <1389908+richvdh@users.noreply.github.com>

hughns committed:
Update proposals/4108-oidc-qr-login.md

Co-authored-by: Richard van der Hoff <1389908+richvdh@users.noreply.github.com>

hughns committed:
Typo + format

hughns committed:
Remove titles from diagrams

hughns committed:
Fix intro to sequence diagrams

hughns committed:
Minor corrections

hughns committed:
Add section on Sign in with QR remote phishing threat

hughns committed:
Add Malicious client sends arbitrary verification URI security consideration

hughns committed:
Add Device ID confusion threat and PoP alternative

hughns committed:
Add alternative to use Authorization Code Grant

hughns committed:
Make more concise

hughns committed:
Add note on which ACG shape feels most appropriate

hughns committed:
Clarifications around Device ID confusion
